core/src/main/java/com/graphhopper/routing/weighting/custom/PriorityCalculator.java

```diff
@@ -24,112 +24,53 @@
-        add(lookup, customModel.getVehicleHeight(), "vehicle_height", MaxHeight.KEY);
-        add(lookup, customModel.getVehicleLength(), "vehicle_length", MaxLength.KEY);
-
         for (Map.Entry<String, Object> entry : customModel.getPriority().entrySet()) {
             String key = entry.getKey();
+            String priorityKey = "priority." + key;
             Object value = entry.getValue();
-
-            if (value == null) {
+            if (value == null)
                 throw new IllegalArgumentException("Missing value for " + key + " in 'priority'");
-            } else if (value instanceof Number) {
-                if (key.startsWith(GeoToValueEntry.AREA_PREFIX)) {
-                    Geometry geometry = GeoToValueEntry.pickGeometry(customModel, key);
-                    priorityList.add(new GeoToValueEntry(new PreparedGeometryFactory().create(geometry), ((Number) value).doubleValue(), 1));
+
+            if (key.startsWith(GeoToValueEntry.AREA_PREFIX)) {
+                if (!(value instanceof Number))
+                    throw new IllegalArgumentException(priorityKey + ": area entry requires number value but was: " + value.getClass().getSimpleName());
+                Geometry geometry = GeoToValueEntry.pickGeometry(customModel, key);
+                priorityList.add(GeoToValueEntry.create(priorityKey, new PreparedGeometryFactory().create(geometry),
+                        (Number) value, 1, 0, 1));
+            } else {
+                if (!(value instanceof Map))
+                    throw new IllegalArgumentException(priorityKey + ": non-root entries requires a map but was: " + value.getClass().getSimpleName());
+                EncodedValue encodedValue = getEV(lookup, "priority", key);
+                if (encodedValue instanceof EnumEncodedValue) {
+                    priorityList.add(EnumToValueEntry.create(priorityKey, (EnumEncodedValue) encodedValue,
+                            (Map) value, 1, 0, 1));
+                } else if (encodedValue instanceof DecimalEncodedValue) {
+                    priorityList.add(DecimalToValueEntry.create(priorityKey, (DecimalEncodedValue) encodedValue,
+                            (Map) value, 1, 0, 1));
+                } else if (encodedValue instanceof BooleanEncodedValue) {
+                    priorityList.add(BooleanToValueEntry.create(priorityKey, (BooleanEncodedValue) encodedValue,
+                            (Map) value, 1, 0, 1));
+                } else if (encodedValue instanceof IntEncodedValue) {
+                    // TODO NOW
                 } else {
-                    BooleanEncodedValue encodedValue = getEV(lookup, "priority", key, BooleanEncodedValue.class);
-                    priorityList.add(new BooleanToValueEntry(encodedValue, ((Number) value).doubleValue(), 1));
+                    throw new IllegalArgumentException("encoded value class '" + encodedValue.getClass().getSimpleName()
```

Comment:
This could be a bit more user friendly: `The encoded value 'key' used in 'priority' is of type X, but only A,B&C are allowed`?